vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+            if (closest_ec != -1)
+                test_ec.pred.multiclass = b.examples[closest_ec]->l.multi.label;
+            else
+                test_ec.pred.multiclass = 0;
+            
+            if (test_ec.l.multi.label != test_ec.pred.multiclass){
+                test_ec.loss = test_ec.weight;
+                b.num_mistakes++;
+            }
+        }
+        else{
+            float reward = 0.f;
+            closest_ec = pick_nearest(b, base, cn, ec); 
+            if (closest_ec != -1){
+                reward = F1_score_for_two_examples(ec, *b.examples[closest_ec]);
+                b.F1_score += reward;
+            }
+            v_array<uint32_t> selected_labs = v_init<uint32_t>();
+            ec.loss = compute_hamming_loss_via_oas(b, base, cn, ec, selected_labs);
+            test_ec.loss = ec.loss;
+            b.hamming_loss += ec.loss;
+        }
+	free_example(&ec);
+    }
+
+
+
+    float return_reward_from_node(memory_tree& b, single_learner& base, uint32_t cn, example& ec, float weight = 1.f){
+        //example& ec = *b.examples[ec_array_index];
+        MULTICLASS::label_t mc;
+        uint32_t save_multi_pred;
+        MULTILABEL::labels multilabels;
+        MULTILABEL::labels preds;
+        if (b.oas ==false){
+            mc = ec.l.multi;
+            save_multi_pred = ec.pred.multiclass;
+        }
+        else{
+            multilabels = ec.l.multilabels;
+            preds = ec.pred.multilabels;
+        }
+	ec.l.simple = {FLT_MAX, 1., 0.0};
+        while(b.nodes[cn].internal != -1){
+            base.predict(ec, b.nodes[cn].base_router);
+            float prediction = ec.pred.scalar;
+            cn = prediction < 0 ? b.nodes[cn].left : b.nodes[cn].right;
+        }
+        
+        if(b.oas == false){
+            ec.l.multi = mc; 
+            ec.pred.multiclass = save_multi_pred;
+        }
+        else{
+            ec.pred.multilabels = preds;
+            ec.l.multilabels = multilabels;
+        }
+
+        //get to leaf now:
+	    int64_t closest_ec = 0;
+        float reward = 0.f;
+        closest_ec = pick_nearest(b, base, cn, ec); //no randomness for picking example.
+        if (b.oas == false){
+            if ((closest_ec != -1) && (b.examples[closest_ec]->l.multi.label == ec.l.multi.label))
+                reward = 1.f;
+        }
+        else{
+            if (closest_ec != -1)
+                reward = F1_score_for_two_examples(ec, *b.examples[closest_ec]);
+        }
+        b.total_num_queires++;
+
+        if (b.learn_at_leaf == true && closest_ec != -1){
+        	float score = normalized_linear_prod(b, &ec, b.examples[closest_ec]);
+        	example* kprod_ec = &calloc_or_throw<example>();
+        	diag_kronecker_product_test(ec, *b.examples[closest_ec], *kprod_ec, b.oas);
+         	kprod_ec->l.simple = {reward, 1.f, -score};
+	    	kprod_ec->weight = weight;
+            base.learn(*kprod_ec, b.max_routers);
+            free_example(kprod_ec);
```

Comment:
Nice to avoid a create/free cycle.

Reply:
seems hard here?  It is training a scorer that takes two examples and learns a similarity score of them. So here I'm creating a new example whose features are the Kronecker product of the two given examples

Reply:
Maybe allocate this example in memory_tree& b and only free it after all examples are processed?

Reply:
this is doable, but I think it will cause serious memory problem (we will double the memory usage during the learning process) for large datasets..

Reply:
I'm not following.  A single example doesn't cause a serious problem?

Reply:
oh I think I see what you meant....let me work on this...

Reply:
done,